### DISCUSSION TITLE
ESP8266 WIFI

### DISCUSSION THREAD

**Discussion by PatLub85** [micropython/micropython#10618] [General]
Hi,
I need your help because I don't understand what's going on. I have 2 esp8266 , one in AP mode and the second in station mode.
When I connect the ESP in station mode to my box, it's ok.
On the other hand when I use this code to connect to the other ESP in AP mode with a fixed IP, the ESP remains in the while loop
```import network

def config_wifi():
    ip        = '192.168.144.3'
    subnet    = '255.255.255.0'
    gateway   = '192.168.144.1'
    dns       = '8.8.8.8'
    ssid = 'MicroPython-AP'
    password = '123456789'
    network.WLAN(network.AP_IF).active(False) 
    sta=network.WLAN(network.STA_IF)
    sta.active(True)
    sta.disconnect() # for ESP8266
    sta.ifconfig((ip,subnet,gateway,dns))
    sta.connect(ssid,password)
    while (sta.isconnected()==False):
        print(sta.isconnected())
    ip=sta.ifconfig()[0] 
    print('Connected',ip)
    return ip

def run():
    config_wifi()
    
run()
```
I've already had the two ESPs working together, but I don't know what I changed, but for a few days nothing has worked.
I use MicroPython v1.19.1 and Thonny 4.0.2

Can you help me
Thank's

**Comment by davefes** [comment]
It has been a long time since I worked with the ESP8266, however one "feature" of this device is that it starts off in AP.  Maybe, even making the `AP.active(False)` isn't enough. 

For the ESP32 I do a:
```py
 #  seems more reliable to start with a fresh connect()
    if sta_if.isconnected():
           ###sta_if.disconnect()
        print (f'started in the connected state, but now disconnected')
    else:
        print (f'started in the disconnected state')
```
So, I would suggest defining both a STA and an AP interface, disconnecting from both of then set up the interface you want.

Also, with the ESP32 there still seems to be some uncertainty when you can do:
`sta.ifconfig((ip,subnet,gateway,dns))`, ie before or after a connect

Good luck!

**Reply by PatLub85** [reply]
Hi Davefes,
I tried 
```
ap = network.WLAN(network.AP_IF)
ap.active(True)
ap.disconnect()  
station = network.WLAN(network.STA_IF)
station.active(True)
station.disconnect()
```
I have this message OSError: STA required for the line ap.disconnect() 
Isn't possible to deconnect an AP?

**Reply by davefes** [reply]
Sorry you need someone with ESP8266 experience.

What happens if you leave out:
`sta.ifconfig((ip,subnet,gateway,dns))`

Also, try the `sta.ifconfig((ip,subnet,gateway,dns))`  after you connect like you did with
`ip=sta.ifconfig()[0]`

**Reply by PatLub85** [reply]
Following my tests:
I used an ESP32 Lolin with this program and I have the same result, ie False in the while loop.
```
import network
from time import sleep

def config_wifi():
    ip        = '192.168.144.3'
    subnet    = '255.255.255.0'
    gateway   = '192.168.144.1'
    dns       = '8.8.8.8'
    ssid = 'MicroPython-AP'
    password = '123456789'
    ap=network.WLAN(network.AP_IF)
    ap.active(False)
    ap.disconnect()  
    print(ap.status()) # return None
    sleep(2)
    sta=network.WLAN(network.STA_IF)
    sta.active(True)
    sta.disconnect() # for ESP8266
    sta.connect(ssid,password)
    print(sta.status())# return 1001 : that is to say CONNEXION
    sleep(2) 
    while (sta.isconnected()==False):
        print(sta.isconnected())
    sta.ifconfig((ip,subnet,gateway,dns))
    ip=sta.ifconfig()[0] 
    print('Connected',ip)
    return ip

def run():
    config_wifi()
    
run()
```
The first status gives me None and the second 1001.
Searching the forum, 1001 means Connexion.
Status indicates that there is connection, on the other hand the while loop gives False
I do not understand

**Reply by davefes** [reply]
```py
    sta.connect(ssid,password)
    print(sta.status())# return 1001 : that is to say CONNEXION
    sleep(2) 
    while (sta.isconnected()==False):
        print(sta.isconnected())
```
I would change to:
```py
    sta.connect(ssid,password)
    while not sta.isconnected():
        print(sta.isconnected())
    print(sta.status())
```
connect I believe is blocking so maybe that first `print(sta.status)`is untrue??
I would remove that sleep(2) and let the following:
```py
while not sta.isconnected():
    print(sta.isconnected())
```
loop around until the connection is really established.

**Comment by PatLub85** [comment]
Hi,
I still have the problem. Now the second status is 1.
I flashed the ESP again and if sta.ifconfig((ip,subnet,gateway,dns)) is executed, afterwards It keeps in memory the parameters of the previous connections.
Is there a way to clear this information from memory?

**Reply by davefes** [reply]
> I flashed the ESP again

An ESP8266, correct?  What I do sometimes on my ESP32s is a power-on reset.  I recall that that might be a bit more tricky for the ESP8266.  What happens if you remove power from the board  ... is that previous information still retained.

**Comment by PatLub85** [comment]
Hi,
I didn't do anything special and now it works.
These little machines are very likely
